### DISCUSSION TITLE
Watchers and modules vs resource efficiency

### DISCUSSION THREAD

**Discussion by tigersoul925** [Hammerspoon/hammerspoon#3437] [Q&A]
I've recently moved toward putting things into separate modules. This is helpful for readability and keeping things separate, but the fun ends when the need arise for watchers that are shared between several modules.

The natural thing to do is adding a seperate watcher object to every module, but will I use up extra resources this way rather than having a single watcher object and sending signals to modules (or however I would resolve it)?

**Comment by cmsj** [comment]
Yep, this is one of the most significant architectural shortcomings in Hammerspoon, and frankly is one of the big reasons why I've been reluctant to just call a 1.0 release.

What I would like to do is rewrite all of the modules that have watchers to use http://www.hammerspoon.org/docs/hs.watchable.html in such a way that the watchers are started and stopped automatically depending on whether there are any subscribers.

I don't know if we'll ever actually get to that though, and if we did, even though we're <1.0 I would expect to get a lot of whinging about broken configs!

**Reply by asmagill** [reply]
Basically what I was going to suggest as well -- using `hs.watchable`. Don't have any good examples at the moment, but that was the premise behind the module -- to have a single place where something could change and automatically notify subscribers of said change.

**Reply by tigersoul925** [reply]
Cool, didn't know about hs.watchable. Seems a great way to exchange data between modules when things become connected in detailed ways.

However, to be more specific in my question, say I have two modules that need the caffeinate.watcher to tell when the computer goes in and out of sleep. Let's assume they don't need to talk to each other in any other way at the moment. Would invoking an instance of the caffeinate.watcher in both modules be using extra resources compared to combining the two modules into one and only need one caffeinate.watcher?

E.g how efficient are the watchers under the hood? Does it matter if I have multiple instances of the same watcher or is it irrelevant resource usage wise?

@cmsj

**Reply by megalithic** [reply]
I've used hs.watchable for that sort of thing for dock/undock states for my laptop.. 

the "status" is determined in this module:
https://github.com/megalithic/dotfiles/blob/main/config/hammerspoon/lib/watchers/status.lua#L1

the "watchers" are handled here:
https://github.com/megalithic/dotfiles/blob/main/config/hammerspoon/lib/watchers/dock.lua#L114

I'd love to clean this up even further, but this might be a helpful example to build upon.

Cheers! 

HUGE thanks to @cmsj and @asmagill for their constant help in this community!

**Reply by megalithic** [reply]
Also, of note, I too have seen pains with application/window watching, when you have multiple watchers running. I hadn't considered utilizing hs.watchable for those scenarios, I'd only been using them for hardware-related watching. That's a brilliant idea!

**Reply by megalithic** [reply]
quick update; i decided to swap over my disparate application watchers to watching a single watchable instance that is doing the application watcher. if nothing else, it at least feels better having a single source of truth that's easily handled across all my modules if need be.

Here's a reference to the changes if interested:
https://github.com/megalithic/dotfiles/commit/faba4f7e041c3d3b9e924dc1bc3b649a731f2490

**Reply by tigersoul925** [reply]
> I've used hs.watchable for that sort of thing for dock/undock states for my laptop..
> 
> the "status" is determined in this module: https://github.com/megalithic/dotfiles/blob/main/config/hammerspoon/lib/watchers/status.lua#L1
> 
> the "watchers" are handled here: https://github.com/megalithic/dotfiles/blob/main/config/hammerspoon/lib/watchers/dock.lua#L114
> 
> I'd love to clean this up even further, but this might be a helpful example to build upon.
> 
> Cheers!
> 
> HUGE thanks to @cmsj and @asmagill for their constant help in this community!

That's a very neat idea. I see the code is a tad above my current understanding of Hammerspoon/Lua, but I'm doing something very similar in a Docking handling setup. Just that it's spagettied into a single init.lua. But that also means it's a darn mess. This is a good idea. Guess I'll get to working on using hs.watchable to centralize all watchers. *sigh* ;)

**Reply by tigersoul925** [reply]
I started playing with this, but I ran into an issue. When using hs.watchable for watchers of various kinds, there is a problem:

If a watcher receives the same value twice in a row, which is fed into a watcher table, the watchers will not be notified the second time since the value technically doesn't change into something else. How could I work around this? I need to be notified of EVERY set, even if it is the same value as the old value.

**Reply by megalithic** [reply]
@tigersoul925: not sure if this helps, but, you might have to enable external changes in the new contructor: https://www.hammerspoon.org/docs/hs.watchable.html#new

Maybe that'll let you watch anytime an update happens, if you do that, maybe you can explicitly call the `.change()` function to "force" a change?

**Reply by tigersoul925** [reply]
> @tigersoul925: not sure if this helps, but, you might have to enable external changes in the new contructor: https://www.hammerspoon.org/docs/hs.watchable.html#new
> 
> Maybe that'll let you watch anytime an update happens, if you do that, maybe you can explicitly call the `.change()` function to "force" a change?

That idea didn't work for me, might be doing something wrong as I get an error trying it.

The code below did work, but I feel it's not the best workaround:


```
sharedEvents = hs.watchable.new("status") -- must be a . in path

caffeinateWatcher = hs.watchable.watch("status.caffeinate", function(watcher, path, key, old, new)
  if old == nil then old = { event = "none" } end
  print("There was an update to " .. path .. " for key " .. key .. " going from " .. old.event .. " to " .. new.event)
end)

function test()
  sharedEvents.caffeinate = { event = "systemWillSleep", time = hs.timer.secondsSinceEpoch() }
  sharedEvents.caffeinate = { event = "systemWillSleep", time = hs.timer.secondsSinceEpoch() }
  sharedEvents.caffeinate = { event = "systemWillSleep", time = hs.timer.secondsSinceEpoch() }
end
```

**Reply by megalithic** [reply]
I'd have to defer to @cmsj or @asmagill for this one! I'll be keeping a watch on it, too though, as I imagine I'd maybe get bitten by this, too, at some point.

**Comment by asmagill** [comment]
See #3440 

As an example of its use in the console: *edit: stupid it's...*

~~~lua
-- for this example, note that I have a function `finspect` that takes an arbitrary number
-- of arguments and runs `hs.inspect({ ... }, { newline = " ", indent = "" })` on the list

> obj = hs.watchable.new("test")
w1 = hs.watchable.watch("test.key", function(...) print("w1", finspect(...)) end)
w2 = hs.watchable.watch("test.key", function(...) print("w2", finspect(...)) end)

-- Loading extension: watchable


> obj.key = "hello"
w1	{ { -- hs.watchable.watcher for path test.key _active = true, _alwaysNotify = false, _callback = <function 1>, _objKey = "key", _objPath = "test", _path = "test.key" }, "test", "key", nil, "hello" }
w2	{ { -- hs.watchable.watcher for path test.key _active = true, _alwaysNotify = false, _callback = <function 1>, _objKey = "key", _objPath = "test", _path = "test.key" }, "test", "key", nil, "hello" }


> obj.key = "hello"


> w1:alwaysNotify(true)
hs.watchable.watcher for path test.key

> obj.key = "hello"
w1	{ { -- hs.watchable.watcher for path test.key _active = true, _alwaysNotify = true, _callback = <function 1>, _objKey = "key", _objPath = "test", _path = "test.key" }, "test", "key", "hello", "hello" }


> obj.key = "hello"
w1	{ { -- hs.watchable.watcher for path test.key _active = true, _alwaysNotify = true, _callback = <function 1>, _objKey = "key", _objPath = "test", _path = "test.key" }, "test", "key", "hello", "hello" }


> obj.key = "goober"
w1	{ { -- hs.watchable.watcher for path test.key _active = true, _alwaysNotify = true, _callback = <function 1>, _objKey = "key", _objPath = "test", _path = "test.key" }, "test", "key", "hello", "goober" }
w2	{ { -- hs.watchable.watcher for path test.key _active = true, _alwaysNotify = false, _callback = <function 1>, _objKey = "key", _objPath = "test", _path = "test.key" }, "test", "key", "hello", "goober" }

~~~

**Reply by megalithic** [reply]
TIL about `:alwaysNotify` coming soon! Bravo @asmagill! ❤️

**Reply by tigersoul925** [reply]
2023-03-23 12:33:53: *** ERROR: /Users/tigersoul/.hammerspoon/init.lua:996: attempt to call a nil value (method 'alwaysNotify')

I don't understand from the example how alwaysNotify comes into existence 🤔

**Reply by megalithic** [reply]
From the PR he created to add it. It’s not merged yet. 😁

**Reply by tigersoul925** [reply]
> From the PR he created to add it. It’s not merged yet. 😁

Shit, how didn't I see that link? *smacks head into desk* Awesome, changes coming then! great!!

**Reply by tigersoul925** [reply]
Tested alwaysNotify by injecting updated watchable.lua - works beautifully!

**Reply by tigersoul925** [reply]
A minor remark that I noticed is that hs.watchable interacts with my global function named "error". I should probably not name a function that way, but if anyone does, it clashes with hs.watchable for some reason.

**Reply by asmagill** [reply]
Because `error` is a lua built-in command (see http://www.lua.org/manual/5.4/manual.html#pdf-error).

It's likely that you'll have issues with other lua-based modules or spoons as well at some point.

**Reply by tigersoul925** [reply]
> Because `error` is a lua built-in command (see http://www.lua.org/manual/5.4/manual.html#pdf-error).
> 
> It's likely that you'll have issues with other lua-based modules or spoons as well at some point.

Oh crap. I've got some fixing to do. Thanks.
